File: 프로그래머스/unrated/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> arr) {
    int count = 0;
    int sign = 0;
    
    while(sign != 1)
    {
        vector<int> memo(arr.begin(), arr.end());

        for (int i = 0; i < arr.size(); i++)
        {
            if (arr[i] >= 50 && !(arr[i] & 1))
                arr[i] /= 2;
            else if (arr[i] <= 50 && arr[i] & 1)
            {
                arr[i] *= 2;
                arr[i]++;
            }
        }
        count++;
        
        for (int i = 0; i < arr.size(); i++)
        {
            if (arr[i] == memo[i])
                sign = 1;
            else
            {
                sign = 0;
                break;
            }
                
        }
    }
    

    
    return count - 1;
}

File: 프로그래머스/0/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.java
import java.util.Arrays;

class Solution {
    private static void colatz(final int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] >= 50 && arr[i] % 2 == 0) {
                arr[i] /= 2;
            } else if (arr[i] < 50 && arr[i] % 2 == 1) {
                arr[i] *= 2;
                arr[i]++;
            }
        }
    }

    private static boolean check(final int[] arr, final int[] newArr) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != newArr[i]) return false;
        }
        return true;
    }

    public int solution(int[] arr) {
        int answer = 0;
        while (true) {
            int[] newArr = Arrays.copyOf(arr, arr.length);

            colatz(arr);

            if (check(arr, newArr)) return answer;
            answer++;
        }
    }
}